File: backend/app/schemas/progress.py
"""Pydantic schemas for lesson completion progress and goal pace tracking."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel


class ProgressToggle(BaseModel):
    completed: bool


class LessonProgressResponse(BaseModel):
    lesson_id: uuid.UUID
    completed: bool
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class SectionProgressResponse(BaseModel):
    section_id: uuid.UUID
    title: str
    total_lessons: int
    completed_lessons: int


class GoalResponse(BaseModel):
    goal_date: date
    pace_status: str  # "on_track", "ahead", "behind", "completed", "overdue"
    lessons_per_week_needed: float
    days_remaining: int
    completed_early_by_days: int | None = None  # set when completed before goal


class CourseProgressResponse(BaseModel):
    course_id: uuid.UUID
    total_lessons: int
    completed_lessons: int
    percentage: float
    sections: list[SectionProgressResponse]
    lesson_progress: dict[str, bool]  # lesson_id -> completed
    goal: GoalResponse | None = None

File: frontend/src/types/progress.ts
/**
 * TypeScript types for lesson/course progress tracking and goals.
 */
export interface ProgressToggle {
  completed: boolean;
}

export interface LessonProgressResponse {
  lesson_id: string;
  completed: boolean;
  completed_at: string | null;
}

export interface SectionProgressResponse {
  section_id: string;
  title: string;
  total_lessons: number;
  completed_lessons: number;
}

export interface GoalResponse {
  goal_date: string;
  pace_status: "on_track" | "ahead" | "behind" | "completed" | "overdue";
  lessons_per_week_needed: number;
  days_remaining: number;
  completed_early_by_days: number | null;
}

export interface CourseProgressResponse {
  course_id: string;
  total_lessons: number;
  completed_lessons: number;
  percentage: number;
  sections: SectionProgressResponse[];
  lesson_progress: Record<string, boolean>;
  goal: GoalResponse | null;
}

export interface CourseGoalResponse {
  course_id: string;
  course_title: string;
  goal_date: string | null;
  total_lessons: number;
  completed_lessons: number;
  percentage: number;
  pace_status: string | null;
  lessons_per_week_needed: number | null;
  days_remaining: number | null;
  completed_early_by_days: number | null;
}
